consensus/event_loop.go

```diff
@@ -0,0 +1,854 @@
+package consensus
+
+import (
+	"time"
+
+	"github.com/holiman/uint256"
+	"github.com/pkg/errors"
+
+	"github.com/deso-protocol/core/bls"
+	"github.com/deso-protocol/core/collections"
+	"github.com/deso-protocol/core/collections/bitset"
+)
+
+func NewFastHotStuffEventLoop() *FastHotStuffEventLoop {
+	return &FastHotStuffEventLoop{
+		status:          eventLoopStatusNotInitialized,
+		crankTimerTask:  NewScheduledTask[uint64](),
+		nextTimeoutTask: NewScheduledTask[uint64](),
+	}
+}
+
+// Initializes the consensus event loop with the latest known valid block in the blockchain, and
+// the validator set for the next block height. The functions expects the following for the input
+// params:
+//   - crankTimerInterval: crank timer interval duration must be > 0
+//   - timeoutBaseDuration: timeout base duration must be > 0
+//   - tip: the current tip of the blockchain, with the validator set at that block height. This may
+//     be a committed or uncommitted block.
+//   - safeBlocks: an unordered slice of blocks including the committed tip, the uncommitted tip,
+//     all ancestors of the uncommitted tip that are safe to extend from, and all blocks from forks
+//     that are safe to extend from. This function does not validate the collection of blocks. It
+//     expects the server to know and decide what blocks are safe to extend from.
+//
+// Given the above, This function updates the tip internally, stores the safe blocks, and re-initializes
+// all internal data structures that are used to track incoming votes and timeout messages for QC construction.
+func (fc *FastHotStuffEventLoop) Init(
+	crankTimerInterval time.Duration,
+	timeoutBaseDuration time.Duration,
+	tip BlockWithValidators,
+	safeBlocks []BlockWithValidators,
+) error {
+	// Grab the event loop's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the event loop is not already running
+	if fc.status == eventLoopStatusRunning {
+		return errors.New("FastHotStuffEventLoop.Init: event loop is already running")
+	}
+
+	// Validate the scheduled task durations
+	if crankTimerInterval <= 0 {
+		return errors.New("FastHotStuffEventLoop.Init: Crank timer interval must be > 0")
+	}
+	if timeoutBaseDuration <= 0 {
+		return errors.New("FastHotStuffEventLoop.Init: Timeout base duration must be > 0")
+	}
+
+	// Validate the safe blocks and validator sets, and store them
+	if err := fc.storeBlocks(tip, safeBlocks); err != nil {
+		return errors.Wrap(err, "FastHotStuffEventLoop.Init: ")
+	}
+
+	// We track the current view here so we know which view to time out on later on.
+	fc.currentView = tip.Block.GetView() + 1
+
+	// Reset QC construction status for the current view
+	fc.hasCrankTimerRunForCurrentView = false
+	fc.hasConstructedQCInCurrentView = false
+
+	// Reset all internal data structures for votes and timeouts
+	fc.votesSeen = make(map[[32]byte]map[string]VoteMessage)
```

Comment:
Nit: Whenever I see `[32]byte` I suspect maybe we should define a type for it. In this case you maybe want something like `VoteHash` or something.